File: backend/app/schemas/order.py
from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field, field_validator, model_validator

from app.models.order import OrderStatus


class OrderItemCreate(BaseModel):
    menu_item_id: int
    quantity: int = Field(ge=1, le=50)
    option_ids: list[int] = Field(default_factory=list)


class OrderCreate(BaseModel):
    restaurant_id: int
    address: str | None = Field(default=None, min_length=3, max_length=300)
    qr_token: str | None = Field(default=None, max_length=200)
    channel: str | None = Field(default=None, pattern="^(delivery|pickup)$")
    dest_lat: float | None = Field(default=None, ge=-90, le=90)
    dest_lng: float | None = Field(default=None, ge=-180, le=180)
    comment: str | None = Field(default=None, max_length=500)
    pay_method: str = Field(default="cash", pattern="^(cash|online)$")
    promo_code: str | None = Field(default=None, max_length=24)
    scheduled_for: datetime | None = None
    items: list[OrderItemCreate] = Field(min_length=1)

    @model_validator(mode="after")
    def destination(self) -> "OrderCreate":
        if (self.qr_token or "").strip():
            return self
        if self.channel == "pickup":
            return self
        if not self.address:
            raise ValueError("Provide a delivery address, pickup, or a table QR token")
        return self


class OrderItemOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    menu_item_id: int
    name: str
    price: float
    quantity: int
    modifiers: list[dict] = Field(default_factory=list)

    @field_validator("modifiers", mode="before")
    @classmethod
    def _mods(cls, value: object) -> object:
        return value or []


class UserBrief(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    name: str
    phone: str | None


class OrderOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    restaurant_id: int
    customer: UserBrief
    courier: UserBrief | None = None
    restaurant_name: str
    status: OrderStatus
    address: str
    comment: str | None
    subtotal: float
    delivery_fee: float
    total: float
    rating: int | None
    channel: str = "delivery"
    table_object_id: int | None = None
    dest_lat: float | None = None
    dest_lng: float | None = None
    pickup_lat: float | None = None
    pickup_lng: float | None = None
    courier_lat: float | None = None
    courier_lng: float | None = None
    courier_heading: float | None = None
    courier_seen_at: datetime | None = None
    pay_method: str = "cash"
    pay_status: str = "unpaid"
    pay_ref: str | None = None
    checkout_url: str | None = None
    scheduled_for: datetime | None = None
    promo_code: str | None = None
    discount: float = 0
    created_at: datetime
    items: list[OrderItemOut]


class OrderStatusUpdate(BaseModel):
    status: OrderStatus


class OrderRate(BaseModel):
    rating: int = Field(ge=1, le=5)

File: mobile/lib/features/orders/domain/order.dart
class OrderModifier {
  const OrderModifier({
    required this.optionId,
    required this.name,
    this.group = '',
    this.price = 0,
  });

  final int optionId;
  final String name;
  final String group;
  final double price;

  factory OrderModifier.fromJson(Map<String, dynamic> json) => OrderModifier(
    optionId: json['option_id'] as int? ?? 0,
    name: json['name'] as String? ?? '',
    group: json['group'] as String? ?? '',
    price: (json['price'] as num?)?.toDouble() ?? 0,
  );
}

class OrderItem {
  const OrderItem({
    required this.menuItemId,
    required this.name,
    required this.price,
    required this.quantity,
    this.modifiers = const [],
  });

  final int menuItemId;
  final String name;
  final double price;
  final int quantity;
  final List<OrderModifier> modifiers;

  List<int> get optionIds => [for (final m in modifiers) m.optionId];

  String get extrasLabel =>
      [for (final m in modifiers) m.name].where((n) => n.isNotEmpty).join(', ');

  String get ticketLine {
    final extras = extrasLabel;
    return extras.isEmpty ? '$quantity× $name' : '$quantity× $name ($extras)';
  }

  factory OrderItem.fromJson(Map<String, dynamic> json) => OrderItem(
    menuItemId: json['menu_item_id'] as int? ?? 0,
    name: json['name'] as String,
    price: (json['price'] as num).toDouble(),
    quantity: json['quantity'] as int,
    modifiers: [
      for (final m in json['modifiers'] as List<dynamic>? ?? [])
        OrderModifier.fromJson(m as Map<String, dynamic>),
    ],
  );
}

class UserBrief {
  const UserBrief({required this.id, required this.name, this.phone});

  final int id;
  final String name;
  final String? phone;

  factory UserBrief.fromJson(Map<String, dynamic> json) => UserBrief(
    id: json['id'] as int,
    name: json['name'] as String,
    phone: json['phone'] as String?,
  );
}

enum OrderStatus {
  pending,
  confirmed,
  preparing,
  onTheWay('on_the_way'),
  delivered,
  cancelled;

  const OrderStatus([this._wire]);
  final String? _wire;

  String get wire => _wire ?? name;

  static OrderStatus parse(String s) =>
      values.firstWhere((v) => v.wire == s, orElse: () => pending);

  bool get canCancel => this == pending || this == confirmed;

  /// Admin transitions (mirrors backend state machine).
  List<OrderStatus> get adminNext => switch (this) {
    pending => [confirmed, cancelled],
    confirmed => [preparing, cancelled],
    preparing => [onTheWay],
    onTheWay => [delivered],
    _ => [],
  };
  bool get isFinal => this == delivered || this == cancelled;

  bool get canReorder => isFinal;

  /// Next step a courier can push this order to, or null.
  OrderStatus? get courierNext => switch (this) {
    confirmed => preparing,
    preparing => onTheWay,
    onTheWay => delivered,
    _ => null,
  };
}

class Order {
  const Order({
    required this.id,
    required this.restaurantId,
    required this.status,
    required this.address,
    required this.subtotal,
    required this.deliveryFee,
    required this.total,
    required this.createdAt,
    required this.items,
    required this.customer,
    required this.restaurantName,
    this.comment,
    this.courier,
    this.rating,
    this.channel = 'delivery',
    this.tableObjectId,
    this.destLat,
    this.destLng,
    this.pickupLat,
    this.pickupLng,
    this.courierLat,
    this.courierLng,
    this.courierHeading,
    this.courierSeenAt,
    this.payMethod = 'cash',
    this.payStatus = 'unpaid',
  });

  final int id;
  final int restaurantId;
  final String restaurantName;
  final int? rating;
  final UserBrief customer;
  final UserBrief? courier;
  final OrderStatus status;
  final String address;
  final String? comment;
  final double subtotal;
  final double deliveryFee;
  final double total;
  final DateTime createdAt;
  final List<OrderItem> items;
  final String channel;
  final int? tableObjectId;
  final double? destLat;
  final double? destLng;
  final double? pickupLat;
  final double? pickupLng;
  final double? courierLat;
  final double? courierLng;
  final double? courierHeading;
  final DateTime? courierSeenAt;
  final String payMethod;
  final String payStatus;

  bool get isCash => payMethod == 'cash';
  bool get isPaid => payStatus == 'paid';

  bool get isDelivery => channel == 'delivery';
  bool get isPickup => channel == 'pickup';
  bool get isDineIn => channel == 'qr_table';
  bool get hasMap => destLat != null || pickupLat != null || courierLat != null;

  List<OrderStatus> get kitchenNext => status.adminNext;

  factory Order.fromJson(Map<String, dynamic> json) => Order(
    id: json['id'] as int,
    restaurantId: json['restaurant_id'] as int,
    restaurantName: json['restaurant_name'] as String,
    customer: UserBrief.fromJson(json['customer']),
    courier: json['courier'] == null
        ? null
        : UserBrief.fromJson(json['courier']),
    status: OrderStatus.parse(json['status'] as String),
    address: json['address'] as String,
    comment: json['comment'] as String?,
    subtotal: (json['subtotal'] as num).toDouble(),
    deliveryFee: (json['delivery_fee'] as num).toDouble(),
    total: (json['total'] as num).toDouble(),
    createdAt: DateTime.parse(json['created_at'] as String),
    items: (json['items'] as List).map((e) => OrderItem.fromJson(e)).toList(),
    rating: json['rating'] as int?,
    channel: json['channel'] as String? ?? 'delivery',
    tableObjectId: json['table_object_id'] as int?,
    destLat: _coord(json['dest_lat']),
    destLng: _coord(json['dest_lng']),
    pickupLat: _coord(json['pickup_lat']),
    pickupLng: _coord(json['pickup_lng']),
    courierLat: _coord(json['courier_lat']),
    courierLng: _coord(json['courier_lng']),
    courierHeading: _coord(json['courier_heading']),
    courierSeenAt: json['courier_seen_at'] == null
        ? null
        : DateTime.tryParse(json['courier_seen_at'] as String),
    payMethod: json['pay_method'] as String? ?? 'cash',
    payStatus: json['pay_status'] as String? ?? 'unpaid',
  );
}

double? _coord(dynamic value) => value is num ? value.toDouble() : null;
